### PR TITLE
feat(poller): poller modified to be able to run on a differen core

### PR DESCRIPTION
**Pull request by dsavitskiy**
Signed-off-by: Dmitry Savitskiy <dmitry.savitskiy@datacore.com>

### CODE REVIEW

File: src/poller.rs [hunk 1]
```diff
@@ -16,80 +20,317 @@ use crate::{
...
+    let p = PollerInner::<T>::from_ctx(PollerContext(ctx));
+    let g = p.lock.lock();
+
+    match p.state {
+        PollerState::Waiting => {
+            p.state = PollerState::Running;
+            (p.poll_fn)(&p.data);
+        }
+        PollerState::Stopped => {
+            drop(g);
```
Comment (tiagolobocastro) [inline src/poller.rs:209]: why do we need to drop the lock for unregister?

File: src/poller.rs [hunk 2]
```diff
@@ -16,80 +20,317 @@ use crate::{
...
+    fn stop(&mut self) {
+        info!(
+            "Stopping poller '{}' on {}",
+            self.dbg_name(),
+            Thread::current_info()
+        );
+
+        let _g = self.lock.lock();
+
+        assert_ne!(self.state, PollerState::Stopped);
```
Comment (tiagolobocastro) [inline src/poller.rs:158]: Is there never a case where stop is called twice or something like that?

File: src/poller.rs [hunk 3]
```diff
@@ -16,80 +20,317 @@ use crate::{
...
+/// Poller structure that allows us to pause, stop, resume periodic tasks.
+///
+/// # Generic Arguments
+///
+/// * `T`: user-defined poller data.
+pub struct Poller<'a, T = ()>
+where
+    T: 'a + Default + Send,
+{
+    inner: Option<Box<PollerInner<'a, T>>>,
```
Comment (tiagolobocastro) [inline src/poller.rs:244]: Must the inner be an option?

**Review by tiagolobocastro — APPROVED** [review]
I wonder if we could setup a core just for admin work, but have it work in interrupt mode.

File: src/poller.rs (hunk 2 above)
Comment (dsavitskiy) [reply src/poller.rs:158]: No way, it is called from Poller's drop only.
Poller::stop() public method accepts self (not &self), and calls drop, essential it is equivalent to drop.

File: src/poller.rs (hunk 1 above)
Comment (dsavitskiy) [reply src/poller.rs:209]: Unregister unregisters the poller and deletes the PollerContext object, so the lock must be dropped.
This poller wrapper logic actually mimics the behavior of SPDK poller internals. Unfortunately, SPDK poller state are private to SPDK, so we cannot rely on them.

File: src/poller.rs (hunk 3 above)
Comment (dsavitskiy) [reply src/poller.rs:244]: It must be created uninitialized, for the threaded pollers. SPDK poller object is allocated from poller's own thread and assigned from there.

**Pull request merged**